Kubelet: add initial client/server container runtime

feiskyer opened the pull request:
Part of #28789. Initial implementation of  kubelet client/server container runtime #17048.

To facilitate code reviewing, this function will be split into a few small PRs:
- Add `KubeGenericRuntimeManager` which conforms to kubelet Runtime interface and add a runtime plugin interface for `KubeGenericRuntimeManager` (this PR)
- fake runtime/manager and unit testing
- Add gRPC implementation and add options `--container-runtime-endpoint` and `--image-service-endpoint` to kubelet
- Garbage collection
- Init containers

This PR contains only first part. Other parts will be sent out after this one is in.

CC @dchen1107 @yujuhong  and @tmrts @yifan-gu 

Also CC @kubernetes/sig-node @kubernetes/sig-rktnetes

resouer commented on pkg/kubelet/kuberuntime/fake_kuberuntime.go:
```diff
@@ -0,0 +1,482 @@
+}
+
+func (r *fakeKubeRuntime) ListContainers(filter *runtimeApi.ContainerFilter) ([]*runtimeApi.Container, error) {
+	r.Lock()
```
maybe rlock is enough? as for other get methods.

feiskyer replied:
@resouer `fakeKubeRuntime` is only for unit testing now. I think its ok with `Mutex`'s `Lock`.

yifan-gu commented:
@feiskyer Could you split the generated code into separate commit? I have a hard time opening up the commit...

euank commented on pkg/kubelet/kuberuntime/utils.go:
```diff
@@ -0,0 +1,349 @@
+	// containerNamePrefix is used to identify the containers on the node managed by kubelet
+	containerNamePrefix = "k8s"
+
+	// Taken from lmctfy https://github.com/google/lmctfy/blob/master/lmctfy/controllers/cpu_controller.cc
```
nit, technically kernel choices, not lmctfy conventions, e.g. min is from https://github.com/torvalds/linux/blob/v4.0/kernel/sched/sched.h#L268

feiskyer replied:
@euank  Thanks for figuring out, maybe remove those comments is better?

feiskyer commented:
@yifan-gu Generated code is part of #17048, need a rebase after #17048 is in. I'm including it because the compiling and testing depends on it.

euank commented on pkg/kubelet/container/runtime.go:
```diff
@@ -233,7 +233,6 @@ const (
 	// This unknown encompasses all the states that we currently don't care.
 	ContainerStateUnknown ContainerState = "unknown"
-	// Not in use yet.
 	ContainerStateCreated ContainerState = "created"
```
Nit, if this is now used I'd prefer if it were ordered at the top to have `unknown` last and indicate how the lifecycle progresses

feiskyer replied:
Good point. Thanks.

euank commented on pkg/kubelet/kubelet.go:
```diff
@@ -402,66 +406,103 @@ func NewMainKubelet(
-		klet.containerRuntime = rktRuntime
-	default:
-		return nil, fmt.Errorf("unsupported container runtime %q specified", containerRuntime)
+	} else {
```
We can save a level of indentation and possibly improve readability by setting `containerRuntime = "remote"` above, removing the else, and moving `NewKubeGenericRuntimeManager` for the remote bit into its own case.

Alternately, if we're not going to be switching entirely based on `containerRuntime` for the construction, we could maybe move to just having an `if..else` ladder that checks if the endpoint is set or the runtime is one of the known ones, and likewise save a level of indentation. I think this isn't as readable as the first suggestion though.

feiskyer replied:
Thanks. I will switch to first method.

euank commented on pkg/kubelet/kuberuntime/fake_kuberuntime.go:
```diff
@@ -0,0 +1,482 @@
+limitations under the License.
+*/
+
+package kuberuntime
```
Can we put the `fake_` files here in their own package that's only referenced from tests (e.g. `kuberuntime/testing`) so that they don't get compiled into the final kubelet binary?

feiskyer replied:
@euank That way will introduce an import recycle.

euank replied:
So move the tests that need them too :)

We could also rename them to `fake_..._test.go` to not have them be compiled in so long as they won't be used from any other packages.

I know the fake kubelet is called from non-test code and is part of kubelet.go so we have a precedent for this, but imo it's a mistake and we should do *something* to make it so that doesn't happen here.

feiskyer replied:
I'm not sure whether or not they will be used by some other packages. e.g. @dchen1107 once have an idea of make e2e testing over faked container runtime.

resouer replied:
Personally I prefer `runtime` than `kuberuntime` for this pkg, kuberuntime looks like a wrong spelled word. Just personal thinking, btw :).

feiskyer commented:
Splitted the PR into small commits, so it is more easy to review.

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_image.go:
```diff
@@ -0,0 +1,113 @@
+)
+
+// RemoteImage implements kuberuntime.ImageService
+type RemoteImage struct {
```
RemoteImageService sounds like a better name?

feiskyer replied:
How about RemoteImageService?

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_image.go:
```diff
@@ -0,0 +1,113 @@
+	imageClient runtimeApi.ImageServiceClient
+}
+
+// NewRemoteImage creates a new RemoteImage
```
nit: period at the end.

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+	_, err := r.runtimeClient.StopPodSandbox(ctx, &runtimeApi.StopPodSandboxRequest{
+		PodSandboxId: &podSandBoxID,
+	})
+	if err != nil {
```
Just return err?

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+	_, err := r.runtimeClient.DeletePodSandbox(ctx, &runtimeApi.DeletePodSandboxRequest{
+		PodSandboxId: &podSandBoxID,
+	})
+	if err != nil {
```
ditto

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+}
+
+// StopPodSandbox stops the sandbox. If there are any running containers in the
+// sandbox, they should be force terminated.
```
forced to terminate.

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+}
+
+// DeletePodSandbox deletes the sandbox. If there are any running containers in the
+// sandbox, they should be force deleted.
```
be forced to deletion.

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+	return nil
+}
+
+// PodSandboxStatus returns the Status of the PodSandbox.
```
s/Status/status

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+		ContainerId: &rawContainerID,
+		Timeout:     &timeout,
+	})
+	if err != nil {
```
Just return err

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+	_, err := r.runtimeClient.RemoveContainer(ctx, &runtimeApi.RemoveContainerRequest{
+		ContainerId: &rawContainerID,
+	})
+	if err != nil {
```
ditto.

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+	return resp.Status, nil
+}
+
+// Exec execute a command in the container.
```
executes

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+	return resp.Items, nil
+}
+
+// CreateContainer creates a new container in specified PodSandbox
```
period at the end.
the specified.

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+}
+
+// RemoveContainer removes the container. If the container is running, the container
+// should be force removed.
```
forced to removal.

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+	return resp.Status, nil
+}
+
+// ListPodSandbox returns a list of SandBox.
```
sandboxes.

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+
+	go func(stdout, stderr io.WriteCloser) {
+		for {
+			defer close(errChanOut)
```
This doesn't seem right to me, e.g. the second time the loop tries to reads the stdout/stderr, the defer is called again, which will probably cause a close on a closed channel later.

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+	if stdin != nil {
+		go func(stdin io.Reader) {
+			for {
+				defer close(errChanIn)
```
ditto, shouldn't defer in every iteration.

yifan-gu commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,345 @@
+	select {
+	case err = <-errChanIn:
+		return err
+	case err = <-errChanOut:
```
I think we need to check whether errChanIn and errChanOut returns nil or not.

feiskyer commented:
Rebased to resolve conflicts.

resouer commented on pkg/kubelet/kubelet.go:
```diff
@@ -460,6 +468,42 @@ func NewMainKubelet(
 			return nil, err
 		}
 		klet.containerRuntime = rktRuntime
+	case "remote":
```
Do we assume `remote` runtime will be used for any runtime server like frakti/oci-runtime/xxx ?

feiskyer replied:
Here, `remote` is only used for switch..case, the real container runtime is got by `Version()` API. Check https://github.com/kubernetes/kubernetes/pull/28396/files#diff-8da51ab012ecbdf2b013ecbcf1c37033R114 and https://github.com/kubernetes/kubernetes/pull/28396/files#diff-8da51ab012ecbdf2b013ecbcf1c37033R141

resouer commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -0,0 +1,643 @@
+    //     PodSandboxConfig.LogDirectory = `/var/log/pods/<podUID>/`
+    //     ContainerConfig.LogPath = `containerName_Instance#.log`
+    //
+    // WARNING: Log managment and how kubelet should interface with the
```
typo management

resouer commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -0,0 +1,643 @@
+    // aggregate cpu/memory resources limits of all containers).
+    // Note: On a Linux host, kubelet will create a pod-level cgroup and pass
+    // it as the cgroup parent for the PodSandbox. For some runtimes, this is
+    // sufficent. For others, e.g., hypervisor-based runtimes, explicit
```
typo: sufficient

resouer commented on pkg/kubelet/api/v1alpha1/runtime/api.pb.go:
```diff
@@ -0,0 +1,2867 @@
+	Name *string `protobuf:"bytes,1,opt,name=name" json:"name,omitempty"`
+	// ID of the container.
+	Id *string `protobuf:"bytes,2,opt,name=id" json:"id,omitempty"`
+	// State of the contianer.
```
typo: contianer

resouer commented on pkg/kubelet/kuberuntime/fake_kuberuntime_manager.go:
```diff
@@ -0,0 +1,100 @@
+	return nil, f.err
+}
+
+// fakeRuntimeHelper implementes kubecontainer.RuntimeHelper inter
```
typo: implements

yujuhong commented on pkg/kubelet/kuberuntime/garbage_collect.go:
```diff
@@ -0,0 +1,162 @@
+*/
+
+package kuberuntime
+
```
/cc @ronnielai, who has been working on garbage collection recently.

yujuhong commented on pkg/kubelet/kuberuntime/image.go:
```diff
@@ -0,0 +1,123 @@
+	return len(images) > 0, nil
+}
+
+// Gets all images currently on the machine.
```
nit: Add the method `ListImages` name in the comment to be consistent with other comments.

yujuhong commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,232 @@
+	kubernetesContainerPreStopHandlerLabel         = "io.kubernetes.container.preStopHandler"
+)
+
+type labelledPodSandboxInfo struct {
```
s/labelled/labeled

yujuhong commented:
I find this PR huge and a little bit hard to swallow. Can we at least split the PR into a few?
The grpc implementation for client/server can be a separate PR. The rest of the kuberuntime code can also be split up by not trying to implement everything in one PR. For example, maybe garbage collection can be a separate PR, and labels as well.

resouer commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -0,0 +1,643 @@
+    optional string name = 1;
+    // ID of the container.
+    optional string id = 2;
+    // State of the contianer.
```
typo: contianer

resouer commented on pkg/kubelet/kuberuntime/remote/remote_runtime.go:
```diff
@@ -0,0 +1,334 @@
+}
+
+// NewRemoteRuntimeService creates a new RemoteRuntime.
+func NewRemoteRuntimeService(server string, connectionTimout time.Duration) (*RemoteRuntimeService, error) {
```
Do we need retry here?

feiskyer commented:
Spoke to @yujuhong offline. To facilitate code reviewing, this function will be split into a few small PRs:

- Add KubeGenericRuntimeManager which conforms to kubelet Runtime interface and add a runtime plugin interface for KubeGenericRuntimeManager
- fake runtime/manager and unit testing
- Add gRPC implementation and add options --container-runtime-endpoint and --image-service-endpoint to kubelet
- Garbage collection
- Init containers

And this PR only remains the first part.

yujuhong commented:
> Spoke to @yujuhong offline. To facilitate code reviewing, this function will be split into a few small PRs:
> 
> Add KubeGenericRuntimeManager which conforms to kubelet Runtime interface and add a runtime plugin interface for KubeGenericRuntimeManager
> fake runtime/manager and unit testing
> Add gRPC implementation and add options --container-runtime-endpoint and --image-service-endpoint to kubelet
> Garbage collection
> Init containers
> And this PR only remains the first part.

Will take a look after the PR is updated. Thanks!

feiskyer commented:
@yujuhong already done. only remains the first part now.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1148 @@
+
+	for _, s := range sandboxes {
+		name, namespace, _ := kubecontainer.ParsePodFullName(s.GetName())
+		labelledPodInfo := GetPodSandboxInfoFromLable(s.Labels)
```
nit: s/FromLable/FromLabel (needs to be changed in multiple places).

nit: s/labelledPodInfo/labeledPodInfo
I think it's more commonly spelled with one "l" in US. Since it's shorter, we might as well just change to that.

yujuhong commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,232 @@
+	return podSandboxInfo
+}
+
+func GetContainerInfoFromLabel(labels map[string]string) *labelledContainerInfo {
```
s/FromLabel/FromLabels

yujuhong commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,232 @@
+limitations under the License.
+*/
+
+package kuberuntime
```
Reviewed this file. The code is mostly copied/adapted from dockertools/labels.go 
/cc @Random-Liu, FYI

yujuhong replied:
Do you guys think it's worth moving most of the common code to pkg/kubelet/container to reduce code duplication for the time being?

feiskyer replied:
I don't think so. After we refactoring docker to new runtime interface, there should be only one place (this file) handling labels.

yujuhong replied:
I'm fine with that, but if we don't migrate docker soon, this will be a nightmare to maintain.

yujuhong commented on pkg/kubelet/container/runtime.go:
```diff
@@ -231,12 +231,11 @@ func (id DockerID) ContainerID() ContainerID {
 type ContainerState string
 
 const (
+	ContainerStateCreated ContainerState = "created"
```
Please update  PLEG to map this state to unknown for now so that 1) it doesn't panic, and 2) it doesn't generate an extra "container created" event https://github.com/kubernetes/kubernetes/blob/ef0c9f0c5b8efbba948a0be2c98d9d2e32e0b68c/pkg/kubelet/pleg/generic.go#L86

feiskyer replied:
Done.

yujuhong commented on pkg/kubelet/kuberuntime/image.go:
```diff
@@ -0,0 +1,123 @@
+	return images, nil
+}
+
+// Removes the specified image.
```
nit: Add the full method name in the comment to be consistent with the methods above.

yujuhong commented on pkg/kubelet/kuberuntime/image.go:
```diff
@@ -0,0 +1,123 @@
+	return nil
+}
+
+// Returns Image statistics.
```
nit: Add the full method name in the comment to be consistent with the methods above.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime.go:
```diff
@@ -0,0 +1,73 @@
+/*
```
Move this file to a different package.

feiskyer replied:
Good point. Will do.

feiskyer commented:
@yujuhong Addressed comments.

* Rename labelled to labeled
* Move kuberuntime.go to pkg/kubelet/container/
* Rename ImageService to ImageManagerService because there is already a same name interface inside pkg/kubelet/container/

feiskyer commented:
@k8s-bot test this issue #IGNORE

Random-Liu commented on pkg/kubelet/container/sync_result.go:
```diff
@@ -54,6 +54,7 @@ var (
 	ErrKillContainer    = errors.New("KillContainerError")
 	ErrVerifyNonRoot    = errors.New("VerifyNonRootError")
 	ErrRunInitContainer = errors.New("RunInitContainerError")
+	ErrKillPodSandbox   = errors.New("KillPodSandboxError")
```
@feiskyer The errors and sync results are a bit annoying. :P 
However, since you've already added `ErrKillPodSandbox`, why don't we add `ErrCreatePodSandbox`?

It seems that you just use `err` and `fmt.Sprintf("CreatePodSandbox failed: %v", err)`. https://github.com/kubernetes/kubernetes/pull/28396/commits/6e8f82db7e2c7057544b5dd8bdc9356900bc68ad#diff-8da51ab012ecbdf2b013ecbcf1c37033R406

feiskyer replied:
@Random-Liu Missed this, will add.

Random-Liu commented on pkg/kubelet/kuberuntime/image.go:
```diff
@@ -0,0 +1,123 @@
+	imgSpec := &runtimeApi.ImageSpec{Image: &img}
+	var auth *runtimeApi.AuthConfig
+	if len(creds) > 0 {
+		authConfig := creds[0].AuthConfig
```
Why do you only use the first one here?
In the original implementation, we'll try credentials we get from docker keyring one by one, until success.
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/docker.go#L215-L223

feiskyer replied:
@Random-Liu Will fix.

Random-Liu commented on pkg/kubelet/kuberuntime/image.go:
```diff
@@ -0,0 +1,123 @@
+// secrets if necessary.
+func (m *kubeGenericRuntimeManager) PullImage(image kubecontainer.ImageSpec, pullSecrets []api.Secret) error {
+	img := image.Image
+	repoToPull, _, _, err := parsers.ParseImageName(img)
```
For docker, the `image.Image` may not have tag, so we've to apply default tag ":latest" here, or else docker will pull all tags of the image.
This is not a good place to do such docker specific thing, but we should at least know the potential issue for docker here.

feiskyer replied:
@Random-Liu We may do docker specific things in docker shim and keep general in kubeGenericRuntimeManager.

Random-Liu replied:
@feiskyer Yeah, agree. Just think loudly. :)

feiskyer replied:
@Random-Liu Opened another issue to track docker specific things at #29252.

yujuhong replied:
Thanks for opening the issue!

Random-Liu commented on pkg/kubelet/kuberuntime/image.go:
```diff
@@ -0,0 +1,123 @@
+		}
+	}
+
+	err = m.imageService.PullImage(imgSpec, auth)
```
Track potential issue for docker https://github.com/kubernetes/kubernetes/pull/16580

feiskyer replied:
@Random-Liu Opened another issue to track docker specific things at #29252.

feiskyer commented:
@Random-Liu Addressed comments, PTAL.

Random-Liu commented on pkg/kubelet/kuberuntime/image.go:
```diff
@@ -0,0 +1,123 @@
+}
+
+// Returns Image statistics.
+func (m *kubeGenericRuntimeManager) ImageStats() (*kubecontainer.ImageStats, error) {
```
Just curious, why leave this "not implemented"? Is there any particular reason? :)

feiskyer replied:
There is no image stats api in new runtime interface yet. We may add it in following PRs.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1148 @@
+		keyring:             credentialprovider.NewDockerKeyring(),
+	}
+
+	typedVersion, err := kubeRuntimeManager.runtimeService.Version(kubeRuntimeAPIVersion)
```
@feiskyer Sorry, I don't quite understand. Why we pass in `kubeRuntimeAPIVersion` here?
What is the purpose?

feiskyer replied:
@Random-Liu Kubelet is a client of runtime API, passing `kubeRuntimeAPIVersion` so that container runtimes (actually server of runtime API) can process versioning correctly.

Random-Liu replied:
@feiskyer Hm...Don't quite understand. If so, this should be passed when initializing the client, right?

feiskyer replied:
@Random-Liu Right.

Random-Liu replied:
Still seems wired to me... We should pass the version requirement when we create the client.

feiskyer replied:
@Random-Liu Agree wired compared with other clients, but this is how grpc working.  There is no native way to specify version while creating the grpc client. For example, a runtime service client can be created by calling `NewRuntimeServiceClient(cc *grpc.ClientConn)` at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/runtime/api.pb.go#L2308

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1148 @@
+		typedVersion.GetRuntimeApiVersion())
+
+	// If the container logs directory does not exist, create it.
+	if _, err := osInterface.Stat(containerLogsDir); err != nil {
```
`os.IsNotExist`?

feiskyer replied:
Ack.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+		return nil, err
+	}
+
+	return newRuntimeVersion(typedVersion.GetRuntimeApiVersion())
```
Track potential docker issue: docker apiserver is not semver.
Ref https://github.com/kubernetes/kubernetes/issues/29252

feiskyer replied:
@Random-Liu Added this to #29252.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+func (m *kubeGenericRuntimeManager) generatePodSandboxConfig(pod *api.Pod) (*runtimeApi.PodSandboxConfig, error) {
+	podFullName := kubecontainer.BuildPodFullName(pod.Name, pod.Namespace)
+	podSandboxConfig := &runtimeApi.PodSandboxConfig{
+		Name:        &podFullName,
```
Do we need to randomize the pod name?

yujuhong replied:
Good catch. I think we need that.

feiskyer replied:
We doesn't randomize pod name today, do we need here?

yujuhong replied:
We randomize part of the infra container name, e.g., 
```
k8s_POD.d8dbe16c_nginx-tnrht_default_ed61dd7a-4ac4-11e6-a11c-42010af00002_72e52082
k8s_POD.d8dbe16c_nginx-tnrht_default_ed61dd7a-4ac4-11e6-a11c-42010af00002_abca38be
```

feiskyer replied:
Y, missed this.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	podFullName := kubecontainer.BuildPodFullName(pod.Name, pod.Namespace)
+	podSandboxConfig := &runtimeApi.PodSandboxConfig{
+		Name:        &podFullName,
+		Labels:      newPodLabels(pod),
```
Labels or annotations?
IIRC, annotation is for arbitrary data, label is for select.
To me, grace period or something looks more like annotation. :)

yujuhong replied:
+1 Let's move this to annotations.

feiskyer replied:
So keep PodSandboxConfig.Labels always nil?

yujuhong replied:
Not always nil. Pod uid, name, namespaces, etc should still be labels. However, it's unlikely people will want to group containers by deletion grace period.

feiskyer replied:
Got point, so we need a `newPodAnnotations`

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+		if podCPULimit != 0 || podCPURequest != 0 {
+			cpuResource := &runtimeApi.ResourceRequirements{}
+			if podCPULimit != 0 {
+				cpulimit := float64(podCPULimit) / 1000
```
@yujuhong What is the unit in our runtime interface. It seems that we didn't define it yet.

feiskyer replied:
@Random-Liu Check at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/runtime/api.proto#L112

Random-Liu replied:
Ha, I see. Thanks. :)

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+			}
+		}
+		if podCPULimit != 0 || podCPURequest != 0 {
+			cpuResource := &runtimeApi.ResourceRequirements{}
```
What if we just initialize with &runtimeApi.ResourceRequirements{}, and apply values despite whether it is 0 or not, which seems cleaner to me.

yujuhong commented on pkg/kubelet/kuberuntime/image.go:
```diff
@@ -0,0 +1,126 @@
+
+	creds, ok := keyring.Lookup(repoToPull)
+	if !ok {
+		glog.V(3).Infof("Pulling image %s without credentials", img)
```
nit: s/%s/%q

yujuhong commented on pkg/kubelet/kuberuntime/image.go:
```diff
@@ -0,0 +1,126 @@
+func (m *kubeGenericRuntimeManager) RemoveImage(image kubecontainer.ImageSpec) error {
+	err := m.imageService.RemoveImage(&runtimeApi.ImageSpec{Image: &image.Image})
+	if err != nil {
+		glog.Errorf("Remove image %s failed: %v", image.Image, err)
```
nit: s/%s/%q

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	imageService   kubecontainer.ImageManagerService
+}
+
+var _ kubecontainer.Runtime = &kubeGenericRuntimeManager{}
```
I think this is redundant since `NewKubeGenericRuntimeManager` returns `kubecontainer.Runtime` already

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+const (
+	kubeRuntimeAPIVersion       = "0.1.0"
+	containerLogsDir            = "/var/log/pods"
+	minimumGracePeriodInSeconds = 2
```
nit: add comments for these constants.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+)
+
+type kubeGenericRuntimeManager struct {
+	runtimeType         string
```
What is `runtimeType`? Is it just the name of the container runtime?

feiskyer replied:
Y, it is the name of the runtime, maybe runtimeName is more clear

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+
+const (
+	kubeRuntimeAPIVersion       = "0.1.0"
+	containerLogsDir            = "/var/log/pods"
```
nit: the variable name is the same as `/var/log/containers` in kubelet.go. I know this is in a separate package, but let's try to make it less misleading. How about renaming it to `podLogsRootDirectory`?

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+
+	// If the container logs directory does not exist, create it.
+	if _, err := osInterface.Stat(containerLogsDir); err != nil {
+		if err := osInterface.MkdirAll(containerLogsDir, 0755); err != nil {
```
Add a TODO to move the directory creation to kubelet.go

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+		}
+	}
+
+	if serializeImagePulls {
```
@ronnielai, just FYI, since you are working on merging the two image pullers.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	for _, c := range containers {
+		labelledInfo := GetContainerInfoFromLabels(c.Labels)
+		pod, found := pods[labelledInfo.PodUID]
+		if !found {
```
This could cause trouble. In docker's case, if the infra container died unexpectedly, kubelet still want to see and handle containers in the pod -- mostly to get the status and kill them. Can we still get the pod UID from the container labels and fix this?

feiskyer replied:
Ack.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+
+		converted, err := m.toRuntimeContainer(c)
+		if err != nil {
+			glog.Warningf("Convert %s container %v failed: %v", m.runtimeType, c, err)
```
Would be great to print the pod UID as well.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+		}
+	}
+
+	containers, err := m.getKubeletContainers(all)
```
self note: this function will incur at least two `docker ps -a` operations. One from `getKubeletPodSandboxes` and one from `getKubeletContainers`. This is not ideal, but I don't have any great idea to improve this.

feiskyer replied:
Y, there are also duplicated calls on pod-native container runtimes (e.g. hyper), they could get container info directly from pod info.

feiskyer replied:
@yujuhong Maybe reconsider container list/status as part of PodSandbox?

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+		cgroupParent = opts.CgroupParent
+	}
+
+	if cgroupParent != "" {
```
Should we group all Linux-specific logic to a separate function or does it not make sense at all?

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	podSandboxConfig, err := m.generatePodSandboxConfig(pod)
+	if err != nil {
+		message := fmt.Sprintf("GeneratePodSandboxConfig failed: %v", err)
+		glog.Error(message)
```
In general, we try to use `format.Pod(pod)` in log messages to help identify the pod. Could you change the glog lines to do that?

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	podSandBoxID, err := m.runtimeService.CreatePodSandbox(podSandboxConfig)
+	if err != nil {
+		message := fmt.Sprintf("CreatePodSandbox failed: %v", err)
+		glog.Error(message)
```
same here: https://github.com/kubernetes/kubernetes/pull/28396/files#r71623259

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	return podSandBoxID, nil, ""
+}
+
+func (m *kubeGenericRuntimeManager) makeMounts(containerFullName string, opts *kubecontainer.RunContainerOptions, container *api.Container) []*runtimeApi.Mount {
```
This doesn't need to be a method of `kubeGenericRuntimeManager`. We can move it to another file to keep this file readable.

feiskyer replied:
What about move makeMounts to utils.go?

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	return volumeMounts
+}
+
+func (m *kubeGenericRuntimeManager) makeContainerLogsPath(containerName, podUID string) string {
```
nit: s/makeContainerLogsPath/getContainerLogsPath

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+
+	_, containerName, _ := buildContainerName(pod.Name, pod.Namespace, container)
+	command, args := kubecontainer.ExpandContainerCommandAndArgs(container, opts.Envs)
+	containerLogsPath := m.makeContainerLogsPath(containerName, string(pod.UID))
```
We can't just use `containerName`. There will be collisions. Need to use the container ID for this.

yujuhong replied:
Hm...just noticed that this is the from `buildContainerName`, so it may be ok. I guess we can't avoid this since we need to generate the path *before* creating the container.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	return path.Join(containerLogsDir, podUID, fmt.Sprintf("%s.log", containerName))
+}
+
+func (m *kubeGenericRuntimeManager) generateContainerConfig(container *api.Container, pod *api.Pod, restartCount int) (*runtimeApi.ContainerConfig, error) {
```
can we also make this a helper method (not part of kubeGenericRuntimeManager) and move it to a separate file with the `makeMounts`? This would make this file more readable IMO.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+/*
```
OK. A general suggestion about this file. Let's move the long methods that populating the configs to a separate file in this package. This will make the lifecycle part move obvious in this file.

feiskyer replied:
@yujuhong Y, splitted some codes into container.go and podsandbox.go

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+
+	ref, err := kubecontainer.GenerateContainerRef(pod, container)
+	if err != nil {
+		glog.Errorf("Can't make a ref to pod %v, container %v: '%v'", pod.Name, container.Name, err)
```
nit: again, instead of using `pod.Name`, if you have a `api.Pod` object, just call `format.Pod()` to print relevant information when logging.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	glog.Infof("Generating ref for container %s: %#v", container.Name, ref)
+
+	// For a new container, the RestartCount should be 0
+	var restartCount int = 0
```
nit: just `restartCount := 0`?

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+		return err, "Generate Container Config Failed"
+	}
+
+	podSandboxConfig, err := m.generatePodSandboxConfig(pod)
```
Why can't we generate the `PodSandboxConfig` once and pass it to each container that we need to start?

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	if err != nil {
+		m.recorder.Eventf(ref, api.EventTypeWarning, events.FailedToStartContainer,
+			"Failed to start container with id %v with error: %v", containerID, err)
+		return err, "Start Container Failed"
```
The message format differs from line 585.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	return config, nil
+}
+
+func (m *kubeGenericRuntimeManager) startContainer(podSandboxID string, container *api.Container, pod *api.Pod, podStatus *kubecontainer.PodStatus, pullSecrets []api.Secret) (error, string) {
```
suggestion: add a brief description of this function. E.g.,
```
startContainers starts a container through the following steps.
 * Pull the image
 * Create a container
 * Start the container
 * Run the post start lifecycle hooks (if applicable)

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	if err != nil {
+		glog.Errorf("Can't make a ref to pod %v, container %v: '%v'", pod.Name, container.Name, err)
+	}
+	glog.Infof("Generating ref for container %s: %#v", container.Name, ref)
```
This is probably `glog.V(4).Infof` or higher.

yujuhong commented on pkg/kubelet/kuberuntime/utils.go:
```diff
@@ -0,0 +1,349 @@
+	return result, nil
+}
+
+// getKubeletContainers lists all container or just the running ones.
```
nit: expand the comment.
`getKubeletContainers lists all (or just the running) containers managed by kubelet.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+		return
+	}
+	if !found {
+		glog.V(4).Infof("Can not find sandbox %s, assuming already removed.", podFullName)
```
`format.Pod()`

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	}
+
+	podFullName := kubecontainer.BuildPodFullName(podName, podNamespace)
+	found, podSandboxID, err := m.getPodSandboxID(podFullName, nil)
```
As @Random-Liu mentioned above, `podFullName` is not sufficient to identify a sandbox. We may have already stopped the sandbox and started a new one. Will it make more sense to just use pod UID to filter sandboxes? That's what we are doing with docker today.

feiskyer replied:
@yujuhong Will do.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+}
+
+// If a container is still in backoff, the function will return a brief backoff error and a detailed error message.
+func (m *kubeGenericRuntimeManager) doBackOff(pod *api.Pod, container *api.Container, podStatus *kubecontainer.PodStatus, backOff *flowcontrol.Backoff) (bool, string, error) {
```
this also looks like a helper method that can be moved to another file.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,1150 @@
+	return changes, nil
+}
+
+// Syncs the running pod into the desired pod.
```
s/Syncs/SyncPod syncs

yujuhong commented:
One concern I have with this PR is that the status of the podsandbox is not passed to `SyncPod()`, hence the method has to call the runtime to find out the status. This is not ideal because we don't want to be invoking runtime operations every time kubelet syncs the pod.

What if we add a new field in the [`kubecontainer.PodStatus`](https://github.com/kubernetes/kubernetes/blob/56b9daf50f12fb56c6ab3961f4820f2b35020c02/pkg/kubelet/container/runtime.go#L257) for the sandbox status, so that `SyncPod` can use it directly? Other packages (dockertools and rkt) can simply ignore this field now.

yujuhong commented:
Adding @Random-Liu as a reviewer. He is familiar with the docker manager implementation and has agreed (thanks!) to help review the PR.

feiskyer commented:
@yujuhong @Random-Liu Thanks for reviewing. Addressed comments and added pod sandbox status to `kubecontainer.PodStatus`. PTAL.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,715 @@
+// exited and dead containers (used for garbage collection).
+func (m *kubeGenericRuntimeManager) GetPods(all bool) ([]*kubecontainer.Pod, error) {
+	pods := make(map[kubetypes.UID]*kubecontainer.Pod)
+	var result []*kubecontainer.Pod
```
nit: `result` is never used until the end, move this to the end may make the code easier to read.

Random-Liu commented on pkg/kubelet/kuberuntime/utils.go:
```diff
@@ -0,0 +1,184 @@
+)
+
+// Converts kubecontainer.ContainerState to runtimeApi.ContainerState
+func toRuntimeContainerState(state kubecontainer.ContainerState) runtimeApi.ContainerState {
```
nit: toRuntimeContainerState converts kubecontainer.ContainerState to runtimeApi.ContainerState.

And also the following functions.

Random-Liu commented on pkg/kubelet/kuberuntime/utils.go:
```diff
@@ -0,0 +1,184 @@
+}
+
+// Converts runtimeApi.Container to kubecontainer.Container.
+func (m *kubeGenericRuntimeManager) toRuntimeContainer(c *runtimeApi.Container) (*kubecontainer.Container, error) {
```
toKubeContainer?

Random-Liu commented on pkg/kubelet/kuberuntime/utils.go:
```diff
@@ -0,0 +1,184 @@
+
+// Converts runtimeApi.Container to kubecontainer.Container.
+func (m *kubeGenericRuntimeManager) toRuntimeContainer(c *runtimeApi.Container) (*kubecontainer.Container, error) {
+	if c == nil || c.Id == nil || c.Name == nil || c.ImageRef == nil || c.State == nil {
```
`ImageRef` is checked here, but `Image` is used later. It still could panic.

feiskyer replied:
@Random-Liu `Image` should be checked instead of `ImageRef`.

Random-Liu commented on pkg/kubelet/kuberuntime/utils.go:
```diff
@@ -0,0 +1,184 @@
+	case api.ProtocolUDP:
+		return runtimeApi.Protocol_UDP
+	}
+
```
Print something here or else we don't even know the input is unexpected.
In the original code, we printed:
```
glog.Warningf("Unknown protocol %q: defaulting to TCP", port.Protocol)
```

Random-Liu commented on pkg/kubelet/kuberuntime/container.go:
```diff
@@ -0,0 +1,432 @@
+	memoryLimit := container.Resources.Limits.Memory().Value()
+	cpuRequest := container.Resources.Requests.Cpu()
+	cpuLimit := container.Resources.Limits.Cpu()
+	var cpuShares int64
```
CPUQuota and CPUPeriod is missing, is this intended?

feiskyer replied:
@Random-Liu Missed this accidently, will add.

Random-Liu commented on pkg/kubelet/kuberuntime/utils.go:
```diff
@@ -0,0 +1,184 @@
+/*
```
Reviewed this file. There are so many files, I've to track in this kind of way. :)

yujuhong commented on pkg/kubelet/container/runtime.go:
```diff
@@ -262,6 +263,8 @@ type PodStatus struct {
 	Namespace string
 	// IP of the pod.
 	IP string
+	// Status of the pod sandbox
```
nit: Add a comment explaining that not every runtime uses this yet, so it could be nil (not set).

yujuhong commented on pkg/kubelet/container/runtime.go:
```diff
@@ -262,6 +263,8 @@ type PodStatus struct {
 	// IP of the pod.
 	IP string
+	// Status of the pod sandbox
+	Status *runtimeApi.PodSandboxStatus
```
nit: Let's change it to `SandboxStatus` to be less ambiguous.

yujuhong commented on pkg/kubelet/kuberuntime/utils.go:
```diff
@@ -0,0 +1,218 @@
+
+const (
+	// containerNamePrefix is used to identify the containers on the node managed by kubelet
+	containerNamePrefix = "k8s-container"
```
I think this will break our backward compatibility using docker. We'll need to handle that in kubelet.
/cc @Random-Liu FYI.

Random-Liu replied:
@yujuhong Do we really need to differentiate container prefix and pod sandbox prefix? :)
Maybe just "k8s" is ok?

feiskyer replied:
Thanks. Will change both prefix to "k8s".

yujuhong commented:
I think the new commits didn't address one point that we mentioned earlier: There could be multiple PodSandBoxes for the same pod present on the node. If that happens, we want kubelet to rectify the situation. In order for kubelet to do it, it needs to be able to detect the situation. Perhaps we can turn the sandbox status in `kubecontainer.PodStatus` into a list? WDYT? @feiskyer @Random-Liu 

I think I still need to make another pass at the SyncPod logic in general. @Random-Liu could you make another quick pass on whether the labels/names will work with the docker implementation? Thanks!

Random-Liu commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,264 @@
+	PodTerminationGracePeriod *int64
+}
+
+type labeledContainerInfo struct {
```
```
PodDeletionGracePeriod
PodTerminationGracePeriod
Hash
RestartCount
TerminationMessagePath
PreStopHandler
```
Should be in annotation.

feiskyer replied:
@Random-Liu Moved those fields to annotations.

Random-Liu commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,264 @@
+	"encoding/json"
+	"strconv"
+
+	"fmt"
```
Move "fmt" to the upper group.

Random-Liu commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,264 @@
+// newPodLabels creates pod labels from api.Pod.
+func newPodLabels(pod *api.Pod) map[string]string {
+	labels := map[string]string{}
+	for k, v := range pod.Labels {
```
@yujuhong Just curious. If we put pod label on sandbox, then do we want to put pod annotation on sandbox?

Random-Liu commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,264 @@
+		glog.Errorf("Unable to get %q from annotations %q: %v", kubernetesPodTerminationGracePeriodLabel, annotations, err)
+	}
+
+	// Remove grace period
```
Why remove annotation here?

feiskyer replied:
These two fields are already stored at `podSandboxInfo.PodDeletionGracePeriod` and `podSandboxInfo.PodTerminationGracePeriod`.

Random-Liu replied:
@feiskyer I see. Thanks for your explain. :)

Random-Liu commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,264 @@
+
+// getPodSandboxInfoFromLabels gets labeledPodSandboxInfo from labels.
+func getPodSandboxInfoFromLabels(labels map[string]string) *labeledPodSandboxInfo {
+	return &labeledPodSandboxInfo{
```
We've already put pod labels on pod sandbox, then how can we retrieval those labels?

Random-Liu commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,264 @@
+		Name:         getStringValueFromLabel(labels, types.KubernetesContainerNameLabel),
+		TerminationMessagePath: getStringValueFromLabel(labels, kubernetesContainerTerminationMessagePathLabel),
+	}
+	if containerHash, err := getUint64ValueFromLabel(labels, kubernetesContainerHashLabel); err != nil {
```
Just nit:
```
if containerInfo.Hash, err := getUint64ValueFromLabel(labels, kubernetesContainerHashLabel); err != nil {
	logContainerError(containerInfo, kubernetesContainerHashLabel, err)
}
```

Random-Liu commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,264 @@
+func newPodLabels(pod *api.Pod) map[string]string {
+	labels := map[string]string{}
+	for k, v := range pod.Labels {
+		labels[k] = v
```
@yujuhong Do we want to store pod labels and annotations on sandbox?

At least we don't need them for now.

Random-Liu commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,264 @@
+		return value
+	}
+	// Do not report error, because there should be many old containers without label now.
+	glog.V(3).Infof("Container doesn't have label %s, it may be an old or invalid container", label)
```
The latest labels are added in 1.2 https://github.com/kubernetes/kubernetes/commit/884dc5c9e5ef7df0679e68a6a98c7ac1d8bcffa4#diff-3982f6b6e2fa8c2c6119f1750e33cc1f. The new runtime interface will be in 1.4.

What is our deprecate policy, can we assume that all labels should be there now? And return error if not.

Random-Liu commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,264 @@
+/*
```
Reviewed this file.
The original label code is quite unclear because of backward compatibility handling.
* We may still need to version and group pod/container annotation https://github.com/kubernetes/kubernetes/pull/25326.
* If we can make sure that people will drain node before upgrade, backward compatibility handling will be much easier.
Anyway, those are out of the scope of this PR.

yujuhong commented:
The image puller needs to support [pull `qps` and `burst`](https://github.com/kubernetes/kubernetes/blob/007257a5adac832d276af10cbe0c617e6189f04a/pkg/kubelet/dockertools/docker.go#L120), could you add this to the TODOs?

Random-Liu commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -0,0 +1,264 @@
+	kubernetesContainerRestartCountLabel           = "io.kubernetes.container.restartCount"
+	kubernetesContainerTerminationMessagePathLabel = "io.kubernetes.container.terminationMessagePath"
+	kubernetesContainerPreStopHandlerLabel         = "io.kubernetes.container.preStopHandler"
+)
```
Now we have a temporary label `io.cadvisor.metric.prometheus` consumed by cadvisor for custom metrics.
I'm not sure whether we still need it in the new runtime interface.
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/labels.go#L49

@vishh @timstclair

yujuhong commented:
We need to tackle nvidia gpu support later. It's in the docker implementation:`container.Resources.Limits.NvidiaGPU`

yujuhong commented on pkg/kubelet/kuberuntime/container.go:
```diff
@@ -0,0 +1,440 @@
+			HostPath:       &v.HostPath,
+			ContainerPath:  &v.ContainerPath,
+			Readonly:       &v.ReadOnly,
+			SelinuxRelabel: &v.SELinuxRelabel,
```
Shouldn't we process the SelinuxRelabel only when the pods has selinux enabled?
https://github.com/kubernetes/kubernetes/blob/007257a5adac832d276af10cbe0c617e6189f04a/pkg/kubelet/dockertools/docker_manager.go#L481

yujuhong commented on pkg/kubelet/kuberuntime/container.go:
```diff
@@ -0,0 +1,440 @@
+		})
+	}
+
+	if opts.PodContainerDir != "" && len(container.TerminationMessagePath) != 0 {
```
Could you also copy the original comments if they still apply here? E.g.,
https://github.com/kubernetes/kubernetes/blob/007257a5adac832d276af10cbe0c617e6189f04a/pkg/kubelet/dockertools/docker_manager.go#L611
https://github.com/kubernetes/kubernetes/blob/007257a5adac832d276af10cbe0c617e6189f04a/pkg/kubelet/dockertools/docker_manager.go#L618

Random-Liu commented on pkg/kubelet/kuberuntime/podsandbox.go:
```diff
@@ -0,0 +1,272 @@
+	var resp []*runtimeApi.PodSandbox
+	var err error
+
+	if all {
```
nit: Why not just use `if-else` to set filter?

Random-Liu commented on pkg/kubelet/kuberuntime/podsandbox.go:
```diff
@@ -0,0 +1,272 @@
+
+	result := []*runtimeApi.PodSandbox{}
+	for _, s := range resp {
+		if s.Name == nil || len(s.GetName()) == 0 {
```
`len(s.GetName()) == 0` should have include `s.Nmae == nil`.

Random-Liu commented on pkg/kubelet/kuberuntime/podsandbox.go:
```diff
@@ -0,0 +1,270 @@
+			continue
+		}
+
+		_, _, _, _, _, err := parseContainerName(s.GetName())
```
nit: Put the logic into a function like `IsPodSandboxManagedByKubelet`, or else the `_, _, _, _, _, err` logic seems a little unclear here. :)

Random-Liu commented on pkg/kubelet/kuberuntime/podsandbox.go:
```diff
@@ -0,0 +1,270 @@
+
+// generatePodSandboxConfig generates pod sandbox config from api.Pod.
+func (m *kubeGenericRuntimeManager) generatePodSandboxConfig(pod *api.Pod) (*runtimeApi.PodSandboxConfig, error) {
+	_, podName, _ := buildContainerName(pod.Name, pod.Namespace, string(pod.UID), nil)
```
nit: If we want to reuse the function. I feel like it would be better to add a new function `buildPodName`, and call `buildContainerName` inside.

Random-Liu commented on pkg/kubelet/kuberuntime/podsandbox.go:
```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		return nil, err
+	}
+	podSandboxConfig.DnsOptions = &runtimeApi.DNSOption{
```
Currently, we won't set dns configuration if the pod is using hostNetwork. This may not work for docker < 1.12 and this is also an ux change.
Ref https://github.com/kubernetes/kubernetes/pull/29378/files.

Random-Liu commented on pkg/kubelet/kuberuntime/podsandbox.go:
```diff
@@ -0,0 +1,270 @@
+		Searches: dnsSearches,
+	}
+
+	hostname, _, err := m.runtimeHelper.GeneratePodHostNameAndDomain(pod)
```
We also don't set hostname if the pod is using hostNetwork.

@yujuhong Should we implement this kind of logic at this level, or specific runtime level?

feiskyer replied:
@Random-Liu We should impl those logic here, but dns and hostname should indeed be omitted if hostNetwork is used. Will make an update on this.

Random-Liu commented on pkg/kubelet/kuberuntime/podsandbox.go:
```diff
@@ -0,0 +1,270 @@
+		return false, err
+	}
+
+	if podStatus.Linux != nil && podStatus.Linux.Namespaces.Options != nil {
```
What if `podStatus.Linux.Namespaces` is `nil`?

Random-Liu commented on pkg/kubelet/kuberuntime/podsandbox.go:
```diff
@@ -0,0 +1,270 @@
+/*
```
Reviewed this file.

Random-Liu commented on pkg/kubelet/api/v1alpha1/runtime/api.pb.go:
```diff
@@ -547,7 +547,7 @@ type PodSandboxConfig struct {
 	//     PodSandboxConfig.LogDirectory = `/var/log/pods/<podUID>/`
```
Reviewed this file.

Random-Liu commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -169,7 +169,7 @@ message PodSandboxConfig {
     //     PodSandboxConfig.LogDirectory = `/var/log/pods/<podUID>/`
```
Reviewed this file.

Random-Liu commented on pkg/kubelet/container/kuberuntime.go:
```diff
@@ -0,0 +1,73 @@
+	// StopPodSandbox stops the sandbox. If there are any running containers in the
+	// sandbox, they should be force terminated.
+	StopPodSandbox(podSandBoxID string) error
+	// DeletePodSandbox deletes the sandbox. If there are any running containers in the
```
nit: s/running containers/containers

Random-Liu commented on pkg/kubelet/container/kuberuntime.go:
```diff
@@ -0,0 +1,73 @@
+/*
```
Reviewed this file.

Random-Liu commented on pkg/kubelet/container/runtime.go:
```diff
@@ -25,6 +25,7 @@ import (
 
```
Reviewed this file.

Random-Liu commented on pkg/kubelet/kuberuntime/container.go:
```diff
@@ -0,0 +1,451 @@
+
+// generateContainerConfig generates container config for kubelet runtime api.
+func (m *kubeGenericRuntimeManager) generateContainerConfig(container *api.Container, pod *api.Pod, restartCount int) (*runtimeApi.ContainerConfig, error) {
+	opts, err := m.runtimeHelper.GenerateRunContainerOptions(pod, container, "")
```
It seems that we never pass ip when `GenerateRunContainerOptions`.
This will cause downward api for pod IP never to work

feiskyer replied:
@Random-Liu Thanks. Fixed.

Random-Liu commented on pkg/kubelet/kuberuntime/container.go:
```diff
@@ -0,0 +1,451 @@
+		Mounts:      makeMounts(cid, opts, container, podHasSELinuxLabel),
+		LogPath:     &containerLogsPath,
+		Stdin:       &container.Stdin,
+		StdinOnce:   &container.StdinOnce,
```
`Tty`?

Random-Liu commented on pkg/kubelet/kuberuntime/container.go:
```diff
@@ -0,0 +1,451 @@
+		Args:        args,
+		WorkingDir:  &container.WorkingDir,
+		Labels:      newContainerLabels(container, pod, restartCount),
+		Annotations: pod.Annotations,
```
@yujuhong Do we want to add pod annotations on container?

@feiskyer Should we also add annotations from `newContainerAnnotations` here?

feiskyer replied:
@Random-Liu will remove those annotations from container

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -0,0 +1,751 @@
+
+// Runs the command in the container of the specified pod using nsenter.
+// Attaches the processes stdin, stdout, and stderr. Optionally uses a
+// tty.
```
Add a TODO here to handle `resize`

yujuhong replied:
issue: #29579

feiskyer commented:
@yujuhong @Random-Liu Rebased and pass podIP to `GenerateRunContainerOptions`. PTAL.

feiskyer committed:
Add fake_kubetime from #29811

feiskyer committed:
Add pkg kuberuntime for client/server container runtime

feiskyer commented:
I'm closing this PR because it is still too huge for reviewing, will split it into small PRs.
